stock_back2back_order_proc/stock.py

```diff
@@ -0,0 +1,552 @@
+#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+#    GNU General Public License for more details.
+#
+#    You should have received a copy of the GNU General Public License
+#    along with this program.  If not, see <http://www.gnu.org/licenses/>.
+#
+##############################################################################
+
+from openerp.osv import fields, orm
+from openerp.tools.translate import _
+from openerp import netsvc
+
+
+class StockLocation(orm.Model):
+    _name = 'stock.location'
+    _inherit = "stock.location"
+    _columns = {
+        'retention_mode': fields.selection(
+            [('retention', _('Retention Mode')), ('thru', _('Thru mode'))],
+            _('Retention Mode'),
+            required=True,
+            help=_("In 'Retention mode' the system wait for the\
+                  whole quantity before the stuff is processed.\n" +
+                   "In 'Thru mode' the shipped quantity is processed regardless\
+                  of the ordered quantity.")
+        ),
+    }
+    _defaults = {
+        'retention_mode': 'retention',
+    }
+
+
+class StockPicking(orm.Model):
+    _name = 'stock.picking'
+    _inherit = "stock.picking"
+
+    def get_move_chain(self, cr, uid, move_id, context=None, move_obj=False):
+        '''Recursively get the chained moves
+        @return list of the chained moves
+        '''
+        if not move_obj:
+            move_obj = self.pool.get('stock.move')
+        move_tbc = move_obj.browse(cr, uid, move_id, context, move_obj)
+
+        # If there is move_dest_id in the chain
+        if move_tbc.move_dest_id:
+            move_chain = self.get_move_chain(
+                cr, uid, move_tbc.move_dest_id.id, context)
+        else:
+            move_chain = []
+
+        move_chain.append(move_tbc)
+        return move_chain
+
+    def copy_pick_chain(self, cr, uid, all_moves, context=None):
+        '''Copy all the picking related to this order.
+
+        @return the dictionary: {old_pick_id: new_pick_id}
+        '''
+        new_picks = {}
+        all_chained_moves = []
+        sequence_obj = self.pool.get('ir.sequence')
+
+        def _get_sequence_name(seq_obj, pick):
+            if pick.type == 'internal':
+                name = seq_obj.get(cr, uid, 'stock.picking')
+            else:
+                name = seq_obj.get(cr, uid, 'stock.picking.' + pick.type)
+            return name
+
+        for move in all_moves:
+            all_chained_moves.extend(
+                self.get_move_chain(cr, uid, move.id, context))
+
+        for move in all_chained_moves:
+            if not move.picking_id:
+                # TODO handle this kind of exception.
+                raise orm.except_orm(
+                    _('Data Error'),
+                    _('There is no picking relates to this move: ') +
+                    '%s' % move.name)
+            if move.picking_id.id and (move.picking_id.id not in new_picks):
```

Comment:
you can get rid of the first part of the test: you've just raised an exception in that case.